pkg/syndesis/action/upgrade.go

```diff
@@ -0,0 +1,194 @@
+			return upgradeCompleted(syndesis, targetVersion)
+		}
+	} else {
+		// Upgrade pod present, checking the status
+		if upgradePod.Status.Phase == v1.PodSucceeded {
+			// Upgrade finished (correctly)
+
+			// Getting the namespace version again for double check
+			newNamespaceVersion, err := version.GetSyndesisVersionFromNamespace(syndesis.Namespace)
+			if err != nil {
+				return err
+			}
+
+			if newNamespaceVersion == targetVersion {
+				logrus.Info("Syndesis resource ", syndesis.Name, " upgraded to version ", targetVersion)
+				return upgradeCompleted(syndesis, targetVersion)
+			} else {
+				logrus.Warn("Upgrade pod terminated successfully but Syndesis version (", newNamespaceVersion, ") does not reflect target version (", targetVersion, ") for resource ", syndesis.Name, ". Forcing upgrade.")
+				target := syndesis.DeepCopy()
+				target.Status.ForceUpgrade = true
+
+				return sdk.Update(target)
+			}
+		} else if upgradePod.Status.Phase == v1.PodFailed {
+			// Upgrade failed
+			logrus.Warn("Failure while upgrading Syndesis resource ", syndesis.Name, " to version ", targetVersion, ": upgrade pod failure")
+
+			target := syndesis.DeepCopy()
+			target.Status.InstallationStatus = v1alpha1.SyndesisInstallationStatusUpgradeFailureBackoff
+			target.Status.Reason = v1alpha1.SyndesisStatusReasonUpgradePodFailed
+			target.Status.LastUpgradeFailure = &metav1.Time{
+				Time: time.Now(),
+			}
+			target.Status.UpgradeAttempts = target.Status.UpgradeAttempts + 1
+
+			return sdk.Update(target)
+		} else {
+			// Still running
+			logrus.Info("Syndesis resource ", syndesis.Name, " is currently being upgraded to version ", targetVersion)
+			return nil
+		}
+
+	}
+
+}
+
+func upgradeCompleted(syndesis *v1alpha1.Syndesis, newVersion string) error {
+	target := syndesis.DeepCopy()
+	target.Status.InstallationStatus = v1alpha1.SyndesisInstallationStatusInstalled
+	target.Status.Reason = v1alpha1.SyndesisStatusReasonMissing
+	target.Status.Version = newVersion
+	target.Status.LastUpgradeFailure = nil
+	target.Status.UpgradeAttempts = 0
+	target.Status.ForceUpgrade = false
+
+	return sdk.Update(target)
+}
+
+func (a *Upgrade) getUpgradeResources(syndesis *v1alpha1.Syndesis) ([]runtime.Object, error) {
+	rawResources, err := syndesistemplate.GetUpgradeResources(syndesis, syndesistemplate.UpgradeParams{
+		InstallParams: syndesistemplate.InstallParams{
+			OAuthClientSecret: "-",
+		},
+		SyndesisVersion: a.operatorVersion,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	resources := make([]runtime.Object, 0)
+	for _, obj := range rawResources {
+		res, err := util.LoadKubernetesResource(obj.Raw)
+		if err != nil {
+			return nil, err
+		}
+		resources = append(resources, res)
+	}
+
+	return resources, nil
+}
+
+func (a *Upgrade) findUpgradePod(resources []runtime.Object) (*v1.Pod, error) {
+	for _, res := range resources {
+		if pod, ok := res.(*v1.Pod); ok {
```

Comment:
This is assuming that only a single pod is present in the upgrade template. That obviously true right now, but maybe it would be better to match on the name, too, instead of matching on the first pod ?